lib/chef/knife/bootstrap/client_builder.rb

```diff
@@ -80,9 +80,12 @@ def node_name
           knife_config[:chef_node_name]
         end
 
-        # @return [String] enviroment from the knife_config
+        # @return [String] enviroment from the knife_config or first_boot_attributes, if none specified by knife_config
         def environment
-          knife_config[:environment]
+          # prefer environment from cli over environment from first-boot attributes
+          knife_config[:environment] or
+            (first_boot_attributes and first_boot_attributes[:environment]) or
+            chef_config[:environment]
```

Comment:
It has a value when the config file you use has an environment key:

- Knife doesn't complain if your knife.rb has an environment setting. The environment key will be picked up in the chef_config object.
- You can pass `-c ~/.chef/client.rb` as a CLI option, and it can read an environment key from that file.